File: jscomp/others/belt_SetString.ml
# 4 "others/belt_Set.cppo.ml"
module I = Belt_internalSetString
# 8 "others/belt_Set.cppo.ml"
module N = Belt_internalAVLset
module A = Belt_Array

type value = I.value
type t = I.t 


let empty = None
let isEmpty = N.isEmpty
let minimum = N.minimum
let minUndefined = N.minUndefined
let maximum = N.maximum
let maxUndefined = N.maxUndefined

let forEach = N.forEach
let forEachU = N.forEachU
let reduce = N.reduce
let reduceU = N.reduceU
let every = N.every
let everyU = N.everyU
let some = N.some                
let someU = N.someU
let keep = N.keepShared              
let keepU = N.keepSharedU
let partition = N.partitionShared
let partitionU = N.partitionSharedU
                   
let size = N.size
let toList = N.toList
let toArray = N.toArray
let fromSortedArrayUnsafe = N.fromSortedArrayUnsafe
let checkInvariantInternal = N.checkInvariantInternal

let rec add  (t : t) (x : value) : t =
  match t with 
    None -> N.singleton x 
  | Some nt  ->
    let v = nt.value in  
    if x = v then t else
      let {N.left = l; right =  r} = nt in 
      if x < v then 
        let ll = add l x in 
        if ll == l then t 
        else N.bal ll v r
      else 
        let rr = add r x in 
        if rr == r then t
        else N.bal l v rr

let mergeMany h arr =   
  let len = A.length arr in 
  let v = ref h in  
  for i = 0 to len - 1 do 
    let key = A.getUnsafe arr i in 
    v .contents<- add v.contents  key 
  done ;
  v.contents 


let rec remove (t : t) (x : value) : t = 
  match t with 
  | None -> t
  | Some n  ->
    let {N.left = l; value = v; right = r} = n in 
    if x = v then 
      match l, r with 
      | None, _ -> r 
      | _, None -> l 
      | _, Some rn -> 
        let v = ref rn.value in 
        let r = N.removeMinAuxWithRef rn v in 
        N.bal l v.contents r
    else
    if x < v then 
      let ll = remove l x in  
      if ll == l then t  
      else N.bal ll v r 
    else 
      let rr = remove r x in 
      if rr == r then t
      else N.bal l v rr
          
let removeMany h arr = 
  let len = A.length arr in 
  let v = ref h in  
  for i = 0 to len - 1 do 
    let key = A.getUnsafe arr i in 
    v .contents<- remove v.contents  key 
  done ;
  v.contents 
          
let fromArray = I.fromArray
let cmp = I.cmp 
let eq = I.eq 
let get = I.get
let getUndefined = I.getUndefined
let getExn = I.getExn                
let subset = I.subset 
let has = I.has

let rec splitAuxNoPivot (n : _ N.node) (x : value) : t * t =   
  let {N.left = l; value = v; right = r} = n in  
  if x = v then l,  r
  else if x < v then
    match l with 
    | None -> 
      None , Some n
    | Some l -> 
      let ll,  rl = splitAuxNoPivot l x in 
      ll,  N.joinShared rl v r
  else
    match r with 
    | None ->
      Some n,  None
    | Some r -> 
      let lr,  rr = splitAuxNoPivot r x in
      N.joinShared l v lr,  rr


let rec splitAuxPivot (n : _ N.node) (x : value) pres : t  * t =   
  let {N.left = l; value = v; right = r} = n in  
  if x = v then begin 
    pres .contents<- true;
    (l, r)
  end
  else if x < v then
    match l with 
    | None -> 
      None, Some n
    | Some l -> 
      let ll,  rl = splitAuxPivot l x pres in 
      ll,  N.joinShared rl v r
  else
    match r with 
    | None ->
      Some n,  None
    | Some r -> 
      let lr,  rr = splitAuxPivot r x pres in
      N.joinShared l v lr,  rr


let split  (t : t) (x : value) =
  match t with 
    None ->
    (None,  None), false
  | Some n  ->    
    let pres = ref false in 
    let v = splitAuxPivot n  x pres  in 
    v, pres.contents

let rec union (s1 : t) (s2 : t) =
  match s1, s2 with
    (None, _) -> s2
  | (_, None) -> s1
  | Some n1, Some n2 (* (Node(l1, v1, r1, h1), Node(l2, v2, r2, h2)) *) ->    
    let h1, h2 = n1.height, n2.height in             
    if h1 >= h2 then
      if h2 = 1 then add  s1 n2.value else begin
        let {N.left = l1; value = v1; right = r1} = n1 in      
        let (l2,  r2) = splitAuxNoPivot n2 v1 in
        N.joinShared (union l1 l2) v1 (union r1 r2)
      end
    else
    if h1 = 1 then add  s2 n1.value else begin
      let {N.left = l2; value = v2; right = r2} = n2 in 
      let (l1, r1) = splitAuxNoPivot n1 v2 in
      N.joinShared (union l1 l2) v2 (union r1 r2)
    end

let  rec intersect (s1 : t) (s2 : t) =
  match s1, s2 with
    (None, _) 
  | (_, None) -> None
  | Some n1, Some n2 (* (Node(l1, v1, r1, _), t2) *) ->
    let {N.left = l1; value = v1; right  = r1} = n1 in  
    let pres = ref false in 
    let l2,r2 =  splitAuxPivot n2 v1 pres in 
    let ll = intersect l1 l2 in 
    let rr = intersect r1 r2 in 
    if pres.contents then N.joinShared ll v1 rr 
    else N.concatShared ll rr 

let rec diff (s1 : t) (s2 : t) =
  match s1, s2 with
  | (None, _) 
  | (_, None) -> s1
  | Some n1, Some n2 (* (Node(l1, v1, r1, _), t2) *) ->
    let {N.left = l1; value = v1; right = r1} = n1 in
    let pres = ref false in 
    let l2, r2 = splitAuxPivot  n2 v1 pres in 
    let ll = diff  l1 l2 in 
    let rr = diff  r1 r2 in 
    if pres.contents then N.concatShared ll rr 
    else N.joinShared ll v1 rr 


    

File: lib/es6/belt_SetString.js


import * as Belt_internalAVLset from "./belt_internalAVLset.js";
import * as Belt_internalSetString from "./belt_internalSetString.js";

function add(t, x) {
  if (t === undefined) {
    return Belt_internalAVLset.singleton(x);
  }
  let v = t.v;
  if (x === v) {
    return t;
  }
  let l = t.l;
  let r = t.r;
  if (x < v) {
    let ll = add(l, x);
    if (ll === l) {
      return t;
    } else {
      return Belt_internalAVLset.bal(ll, v, r);
    }
  }
  let rr = add(r, x);
  if (rr === r) {
    return t;
  } else {
    return Belt_internalAVLset.bal(l, v, rr);
  }
}

function mergeMany(h, arr) {
  let len = arr.length;
  let v = h;
  for (let i = 0; i < len; ++i) {
    let key = arr[i];
    v = add(v, key);
  }
  return v;
}

function remove(t, x) {
  if (t === undefined) {
    return t;
  }
  let v = t.v;
  let l = t.l;
  let r = t.r;
  if (x === v) {
    if (l === undefined) {
      return r;
    }
    if (r === undefined) {
      return l;
    }
    let v$1 = {
      contents: r.v
    };
    let r$1 = Belt_internalAVLset.removeMinAuxWithRef(r, v$1);
    return Belt_internalAVLset.bal(l, v$1.contents, r$1);
  }
  if (x < v) {
    let ll = remove(l, x);
    if (ll === l) {
      return t;
    } else {
      return Belt_internalAVLset.bal(ll, v, r);
    }
  }
  let rr = remove(r, x);
  if (rr === r) {
    return t;
  } else {
    return Belt_internalAVLset.bal(l, v, rr);
  }
}

function removeMany(h, arr) {
  let len = arr.length;
  let v = h;
  for (let i = 0; i < len; ++i) {
    let key = arr[i];
    v = remove(v, key);
  }
  return v;
}

function splitAuxNoPivot(n, x) {
  let v = n.v;
  let l = n.l;
  let r = n.r;
  if (x === v) {
    return [
      l,
      r
    ];
  }
  if (x < v) {
    if (l === undefined) {
      return [
        undefined,
        n
      ];
    }
    let match = splitAuxNoPivot(l, x);
    return [
      match[0],
      Belt_internalAVLset.joinShared(match[1], v, r)
    ];
  }
  if (r === undefined) {
    return [
      n,
      undefined
    ];
  }
  let match$1 = splitAuxNoPivot(r, x);
  return [
    Belt_internalAVLset.joinShared(l, v, match$1[0]),
    match$1[1]
  ];
}

function splitAuxPivot(n, x, pres) {
  let v = n.v;
  let l = n.l;
  let r = n.r;
  if (x === v) {
    pres.contents = true;
    return [
      l,
      r
    ];
  }
  if (x < v) {
    if (l === undefined) {
      return [
        undefined,
        n
      ];
    }
    let match = splitAuxPivot(l, x, pres);
    return [
      match[0],
      Belt_internalAVLset.joinShared(match[1], v, r)
    ];
  }
  if (r === undefined) {
    return [
      n,
      undefined
    ];
  }
  let match$1 = splitAuxPivot(r, x, pres);
  return [
    Belt_internalAVLset.joinShared(l, v, match$1[0]),
    match$1[1]
  ];
}

function split(t, x) {
  if (t === undefined) {
    return [
      [
        undefined,
        undefined
      ],
      false
    ];
  }
  let pres = {
    contents: false
  };
  let v = splitAuxPivot(t, x, pres);
  return [
    v,
    pres.contents
  ];
}

function union(s1, s2) {
  if (s1 === undefined) {
    return s2;
  }
  if (s2 === undefined) {
    return s1;
  }
  let h1 = s1.h;
  let h2 = s2.h;
  if (h1 >= h2) {
    if (h2 === 1) {
      return add(s1, s2.v);
    }
    let v1 = s1.v;
    let l1 = s1.l;
    let r1 = s1.r;
    let match = splitAuxNoPivot(s2, v1);
    return Belt_internalAVLset.joinShared(union(l1, match[0]), v1, union(r1, match[1]));
  }
  if (h1 === 1) {
    return add(s2, s1.v);
  }
  let v2 = s2.v;
  let l2 = s2.l;
  let r2 = s2.r;
  let match$1 = splitAuxNoPivot(s1, v2);
  return Belt_internalAVLset.joinShared(union(match$1[0], l2), v2, union(match$1[1], r2));
}

function intersect(s1, s2) {
  if (s1 === undefined) {
    return;
  }
  if (s2 === undefined) {
    return;
  }
  let v1 = s1.v;
  let l1 = s1.l;
  let r1 = s1.r;
  let pres = {
    contents: false
  };
  let match = splitAuxPivot(s2, v1, pres);
  let ll = intersect(l1, match[0]);
  let rr = intersect(r1, match[1]);
  if (pres.contents) {
    return Belt_internalAVLset.joinShared(ll, v1, rr);
  } else {
    return Belt_internalAVLset.concatShared(ll, rr);
  }
}

function diff(s1, s2) {
  if (s1 === undefined) {
    return s1;
  }
  if (s2 === undefined) {
    return s1;
  }
  let v1 = s1.v;
  let l1 = s1.l;
  let r1 = s1.r;
  let pres = {
    contents: false
  };
  let match = splitAuxPivot(s2, v1, pres);
  let ll = diff(l1, match[0]);
  let rr = diff(r1, match[1]);
  if (pres.contents) {
    return Belt_internalAVLset.concatShared(ll, rr);
  } else {
    return Belt_internalAVLset.joinShared(ll, v1, rr);
  }
}

let empty;

let fromArray = Belt_internalSetString.fromArray;

let fromSortedArrayUnsafe = Belt_internalAVLset.fromSortedArrayUnsafe;

let isEmpty = Belt_internalAVLset.isEmpty;

let has = Belt_internalSetString.has;

let subset = Belt_internalSetString.subset;

let cmp = Belt_internalSetString.cmp;

let eq = Belt_internalSetString.eq;

let forEachU = Belt_internalAVLset.forEach;

let forEach = Belt_internalAVLset.forEach;

let reduceU = Belt_internalAVLset.reduce;

let reduce = Belt_internalAVLset.reduce;

let everyU = Belt_internalAVLset.every;

let every = Belt_internalAVLset.every;

let someU = Belt_internalAVLset.some;

let some = Belt_internalAVLset.some;

let keepU = Belt_internalAVLset.keepShared;

let keep = Belt_internalAVLset.keepShared;

let partitionU = Belt_internalAVLset.partitionShared;

let partition = Belt_internalAVLset.partitionShared;

let size = Belt_internalAVLset.size;

let toList = Belt_internalAVLset.toList;

let toArray = Belt_internalAVLset.toArray;

let minimum = Belt_internalAVLset.minimum;

let minUndefined = Belt_internalAVLset.minUndefined;

let maximum = Belt_internalAVLset.maximum;

let maxUndefined = Belt_internalAVLset.maxUndefined;

let get = Belt_internalSetString.get;

let getUndefined = Belt_internalSetString.getUndefined;

let getExn = Belt_internalSetString.getExn;

let checkInvariantInternal = Belt_internalAVLset.checkInvariantInternal;

export {
  empty,
  fromArray,
  fromSortedArrayUnsafe,
  isEmpty,
  has,
  add,
  mergeMany,
  remove,
  removeMany,
  union,
  intersect,
  diff,
  subset,
  cmp,
  eq,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  keepU,
  keep,
  partitionU,
  partition,
  size,
  toList,
  toArray,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getExn,
  split,
  checkInvariantInternal,
}
/* No side effect */
